File: test_srwma_performance.py
import time
import numpy as np
import my_project

# Create test data
np.random.seed(42)
data = np.random.randn(1_000_000).astype(np.float64)

# Warmup
for _ in range(10):
    _ = my_project.srwma(data, 14)

# Benchmark single run
start = time.perf_counter()
result = my_project.srwma(data, 14)
end = time.perf_counter()
single_time = (end - start) * 1000

print(f"Single run time: {single_time:.3f} ms")

# Check result
print(f"Result shape: {result.shape}")
print(f"First 10 non-NaN values: {result[~np.isnan(result)][:10]}")
print(f"Number of NaN values: {np.sum(np.isnan(result))}")

# Run multiple iterations for better measurement
times = []
for _ in range(100):
    start = time.perf_counter()
    _ = my_project.srwma(data, 14)
    end = time.perf_counter()
    times.append((end - start) * 1000)

print(f"\nMedian time over 100 runs: {np.median(times):.3f} ms")
print(f"Mean time: {np.mean(times):.3f} ms")
print(f"Min time: {np.min(times):.3f} ms")
print(f"Max time: {np.max(times):.3f} ms")

# Test with different kernels
for kernel in ['auto', 'scalar', 'avx2', 'avx512']:
    try:
        times = []
        for _ in range(50):
            start = time.perf_counter()
            _ = my_project.srwma(data, 14, kernel)
            end = time.perf_counter()
            times.append((end - start) * 1000)
        print(f"\nKernel '{kernel}' median time: {np.median(times):.3f} ms")
    except Exception as e:
        print(f"\nKernel '{kernel}' failed: {e}")

File: src/bin/test_srwma_performance.rs
use my_project::indicators::moving_averages::srwma::{srwma_with_kernel, SrwmaInput, SrwmaParams};
use my_project::utilities::enums::Kernel;
use std::time::Instant;

fn main() {
    // Create test data
    let mut data = vec![0.0; 1_000_000];
    let mut seed = 42u64;
    for i in 0..data.len() {
        // Simple linear congruential generator for reproducible random numbers
        seed = seed.wrapping_mul(1664525).wrapping_add(1013904223);
        let rand_val = (seed as f64) / (u64::MAX as f64) * 2.0 - 1.0;
        data[i] = rand_val;
    }
    
    let params = SrwmaParams { period: Some(14) };
    let input = SrwmaInput::from_slice(&data, params);
    
    // Warmup
    for _ in 0..10 {
        let _ = srwma_with_kernel(&input, Kernel::Auto).unwrap();
    }
    
    // Single run
    let start = Instant::now();
    let result = srwma_with_kernel(&input, Kernel::Scalar).unwrap();
    let duration = start.elapsed();
    println!("Single run time (Scalar): {:.3} ms", duration.as_secs_f64() * 1000.0);
    println!("Result length: {}", result.values.len());
    println!("Number of NaN values: {}", result.values.iter().filter(|x| x.is_nan()).count());
    
    // Benchmark different kernels
    for kernel in [Kernel::Scalar, Kernel::Avx2, Kernel::Avx512, Kernel::Auto] {
        let mut times = Vec::new();
        for _ in 0..100 {
            let start = Instant::now();
            let _ = srwma_with_kernel(&input, kernel).unwrap();
            let duration = start.elapsed();
            times.push(duration.as_secs_f64() * 1000.0);
        }
        
        times.sort_by(|a, b| a.partial_cmp(b).unwrap());
        let median = times[times.len() / 2];
        let mean = times.iter().sum::<f64>() / times.len() as f64;
        let min = times[0];
        let max = times[times.len() - 1];
        
        println!("\nKernel {:?}:", kernel);
        println!("  Median: {:.3} ms", median);
        println!("  Mean: {:.3} ms", mean);
        println!("  Min: {:.3} ms", min);
        println!("  Max: {:.3} ms", max);
    }
}